networkx/algorithms/find_euler_path.py

```diff
@@ -0,0 +1,223 @@
+    .. [2] http://en.wikipedia.org/wiki/Eulerian_path
+    .. [3] https://github.com/networkx/networkx/blob/master/networkx/algorithms/euler.py
+    .. [4] https://www.math.ku.edu/~jmartin/courses/math105-F11/Lectures/chapter5-part2.pdf
+    
+    Examples 
+    --------
+    >>> G = nx.Graph([('W', 'N'), ('N', 'E'), ('E', 'W'), ('W', 'S'), ('S', 'E')])
+    >>> list(nx.find_eulerian_path(G))
+    [('W', 'N'), ('N', 'E'), ('E', 'W'), ('W', 'S'), ('S', 'E')]
+
+    >>> G = nx.Digraph([(1, 2), (2, 3)])
+    >>> list(nx.find_eulerian_path(G))
+    [(1,2),(2,3)]
+    """
+    from operator import itemgetter 
+
+    # Verify that graph is connected, short circuit
+    if G.is_directed() and not nx.is_weakly_connected(G):
+        raise nx.NetworkXError("G is not connected.")    
+
+    # is undirected
+    if not G.is_directed() and not nx.is_connected(G):
+        raise nx.NetworkXError("G is not connected.")
+
+    # Now verify if has an eulerian circuit: even condition of all nodes is satified.
+    if nx.is_eulerian(G):
+        x = nx.eulerian_circuit(G) # generator of edges
+        for i in x:
+            yield i
+        
+    # Not all vertex have even degree, check if exactly two vertex have odd degrees.
+    # If yes, then there is an Euler path. If not, raise an error (no euler path can be found)
+    else:
+        g = G.__class__(G)  # copy graph structure (not attributes)
+        
+        # list to check the odd degree condition, and a flag
+        check_odd = []
+        directed = False
+    
+        if g.is_directed():
+            degree = g.in_degree
+            out_degree = g.out_degree
+            edges = g.in_edges_iter
+            get_vertex = itemgetter(0)
+            directed = True
+        else:
+            degree = g.degree
+            edges = g.edges_iter
+            get_vertex = itemgetter(1)
+            
+        # Verify if an euler path can be found. Complexity O(n) ?
+        for vertex in g.nodes():
+            deg = degree(vertex)
+            # directed case
+            if directed:
+                outdeg = out_degree(vertex)
+                if deg != outdeg:
+                    # if we have more than 2 odd nodes, we do a raise (no euler path)
+                    if len(check_odd) > 2:
+                        raise nx.NetworkXError("G doesn't have an Euler Path.")
+                # is odd and we append it. 
+                    else:
+                        check_odd.append(vertex)
+            # undirected case        
+            else:
+                if deg % 2 != 0:
+                    # if we have more than 2 odd nodes, we do a raise (no euler path)
+                    if len(check_odd) > 2:
+                        raise nx.NetworkXError("G doesn't have an Euler Path.")
+                    # is odd and we append it. 
+                    else:
+                        check_odd.append(vertex)
+
+        if directed:
+            def verify_odd_cond(g,check_odd):
```

Comment:
Why is this its own function? Just do the checks directly without defining a new function.